panel/auth.py

```diff
@@ -891,26 +985,94 @@ def set_current_user(self, user):
 
 
+class LogoutHandler(tornado.web.RequestHandler):
+
+    _login_endpoint = '/login'
+
+    _logout_template = LOGOUT_TEMPLATE
 
-class BasicProvider(OAuthProvider):
+    def get(self):
+        self.clear_cookie("user")
+        self.clear_cookie("id_token")
+        self.clear_cookie("access_token")
+        self.clear_cookie(STATE_COOKIE_NAME)
+        html = self._logout_template.render(
+            PANEL_CDN=CDN_DIST,
+            LOGIN_ENDPOINT=self._login_endpoint
+        )
+        self.write(html)
+
+
+class BasicAuthProvider(AuthProvider):
+    """
+    An AuthProvider which serves a simple login and logout page.
+    """
 
-    def __init__(self, basic_login_template=None, logout_template=None):
-        if basic_login_template is None:
+    def __init__(
+        self, login_endpoint=None, logout_endpoint=None,
+        login_template=None, logout_template=None, error_template=None
+    ):
+        if error_template is None:
+            self._error_template = ERROR_TEMPLATE
+        else:
+            with open(error_template) as f:
+                self._error_template = _env.from_string(f.read())
+        if logout_template is None:
+            self._logout_template = LOGOUT_TEMPLATE
+        else:
+            with open(logout_template) as f:
+                self._logout_template = _env.from_string(f.read())
+        if login_template is None:
             self._basic_login_template = BASIC_LOGIN_TEMPLATE
         else:
-            with open(basic_login_template) as f:
+            with open(login_template) as f:
                 self._basic_login_template = _env.from_string(f.read())
-        super().__init__(logout_template=logout_template)
+        self._login_endpoint = login_endpoint or '/login'
+        self._logout_endpoint = logout_endpoint or '/logout'
+        super().__init__()
+
+    @property
+    def get_user(self):
+        def get_user(request_handler):
+            return request_handler.get_secure_cookie("user", max_age_days=config.oauth_expiry)
+        return get_user
 
     @property
     def login_url(self):
-        return '/login'
+        return self._login_endpoint
 
     @property
     def login_handler(self):
-        BasicLoginHandler._basic_login_template = self._basic_login_template
+        BasicLoginHandler._login_endpoint = self._login_endpoint
         return BasicLoginHandler
 
+    @property
+    def logout_url(self):
+        return self._logout_endpoint
+
+    @property
+    def logout_handler(self):
+        if self._logout_template:
+            LogoutHandler._logout_template = self._logout_template
+        LogoutHandler._login_endpoint = self._login_endpoint
+        return LogoutHandler
+
+
+class OAuthProvider(BasicAuthProvider):
+    """
+    An AuthProvider using specific OAuth implementation selected via
+    the global config.oauth_provider configuration.
+    """
+
+    @property
+    def login_handler(self):
+        handler = AUTH_PROVIDERS[config.oauth_provider]
```

Comment:
If we Do:
```
pn.serve(..., oauth_provider="code")
```

We get this error
```
 File "***/serve_app.py", line 135, in serve_app
    pn.serve(
  File "***/panel/io/server.py", line 922, in serve
    return get_server(panels, **kwargs)
  File "***/panel/io/server.py", line 1175, in get_server
    opts['auth_provider'] = provider(
  File "***/panel/auth.py", line 1032, in __init__
    super().__init__()
  File "***/lib/python3.10/site-packages/bokeh/server/auth_provider.py", line 70, in __init__
    self._validate()
  File "***/lib/python3.10/site-packages/bokeh/server/auth_provider.py", line 179, in _validate
    if self.login_handler and self.get_login_url:
  File "***/panel/auth.py", line 1069, in login_handler
    handler = AUTH_PROVIDERS[config.oauth_provider]
KeyError: None
```

This is because you don't set `config.oauth_provider` in `panel/io/server.py` the same way you set it in `panel/command/serve.py`

Reply:
Thanks, fixed.